libslide/src/bignum/add.rs

```diff
@@ -0,0 +1,135 @@
+#![allow(clippy::suspicious_arithmetic_impl)]
+use crate::bignum::Bignum;
+use std::mem;
+use std::ops;
+
+impl ops::Add for Bignum {
+    type Output = Bignum;
+
+    fn add(self, rhs: Bignum) -> Bignum {
+        let is_neg: bool = rhs.is_neg && self.is_neg;
+
+        // this assumes the signs are both positive for now
+        let mut carry: u8 = 0;
+        let mut lhs_size: usize = self.dec.len();
+        let mut rhs_size: usize = rhs.dec.len();
+        let mut res_dec: Vec<u8> = Vec::new();
+
+        // 1. Handle Decimal
+
+        // swap if neccessary to make lhs smaller. Initialize big vec as the bigger vector
+        let big_vec = if lhs_size > rhs_size {
+            mem::swap(&mut lhs_size, &mut rhs_size);
+            self.dec.clone()
+        } else {
+            rhs.dec.clone()
+        };
+
+        // compute the decimal addition up to the size of the small vec
+        let mut res: u8;
+        for i in (0..lhs_size).rev() {
+            res = self.dec[i] + rhs.dec[i] + carry;
+            res_dec.push(res % 10);
+            carry = (res > 9) as u8;
+        }
+
+        // decimals are added to the vec in reverse order so we switch order back
+        res_dec = res_dec.iter().rev().cloned().collect();
+
+        // add the remaining decimals of the larger vector
+        let mut i = lhs_size;
+        while i < rhs_size {
+            res_dec.push((big_vec[i]) % 10);
+            i += 1;
+        }
+
+        // 2. Handle Integer
+
+        lhs_size = self.int.len();
+        rhs_size = rhs.int.len();
+
+        // we reverse the vectors since our carry matters when we handle
+        // adding the larger vector
+        let mut res_int: Vec<u8> = Vec::new();
```

Comment:
No need to create a new vector with no preallocation, just make it the larger of the two existing vectors.

Reply:
You need to make this one a new vector because you need the push_back method for the second part of this logic since it is easier to not store an i. I did switch it for the dec part.

Reply:
I'm not sure I understand that. Why do you need push_back?

Reply:
I think I will rewrite this so I can solve a few of these issues. I don't think we need push_back.

Reply:
Sg